### PR TITLE
Add json container logger

### LINKED ISSUE #1126: Add support for additional log drivers
Currently CRI format is the only one supported. At minimum long term, json-file and journald should also be added. https://github.com/containers/conmon-rs/issues/482 tracks adding splunk as a log driver as well

### PR DESCRIPTION
**Pull request by wasup-yash**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR is to implement additional log driver support for the JSON based log driver which enables the JSON log format.
 
#### Which issue(s) this PR fixes: 
fixes #1126 

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?

```release-note
Added json container logger support.
```

### CODE REVIEW

File: conmon-rs/common/proto/conmon.capnp [hunk 1]
```diff
@@ -58,6 +58,7 @@ interface Conmon {
         enum Type {
             # The CRI logger, requires `path` to be set.
             containerRuntimeInterface @0;
+            jsonInterface @1;
```
Comment (saschagrunert) [inline conmon-rs/common/proto/conmon.capnp:61]:
```suggestion
            json @1;
```

File: conmon-rs/server/src/container_log.rs [hunk 2]
```diff
@@ -16,43 +16,53 @@ pub struct ContainerLog {
 #[derive(Debug)]
 enum LogDriver {
     ContainerRuntimeInterface(CriLogger),
+    JsonInterface(JsonLogger),
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:19]:
```suggestion
    Json(JsonLogger),
```

**Review by saschagrunert — COMMENTED** [review]
Thank you!

Two major points for now:

1. Please rename `JsonInterface` to `Json`, the `Interface` on `ContainerRuntimeInterface` is intentional.
2. You can fix the build by boxing the futures of distinct types. In general, you can use the [FutureExt::boxed](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.boxed) function to convert the different future types into a single common type using trait objects. This means you can apply the following patch to fix the build:

```patch
diff --git a/conmon-rs/server/src/container_log.rs b/conmon-rs/server/src/container_log.rs
index 00363ec8..77bd43af 100644
--- a/conmon-rs/server/src/container_log.rs
+++ b/conmon-rs/server/src/container_log.rs
@@ -2,7 +2,7 @@ use crate::{container_io::Pipe, cri_logger::CriLogger, json_logger::JsonLogger};
 use anyhow::Result;
 use capnp::struct_list::Reader;
 use conmon_common::conmon_capnp::conmon::log_driver::{Owned, Type};
-use futures::future::join_all;
+use futures::{FutureExt, future::join_all};
 use std::sync::Arc;
 use tokio::{io::AsyncBufRead, sync::RwLock};
 
@@ -61,8 +61,8 @@ impl ContainerLog {
             self.drivers
                 .iter_mut()
                 .map(|x| match x {
-                    LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.init(),
-                    LogDriver::JsonInterface(ref mut json_logger) => json_logger.init(),
+                    LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.init().boxed(),
+                    LogDriver::JsonInterface(ref mut json_logger) => json_logger.init().boxed(),
                 })
                 .collect::<Vec<_>>(),
         )
@@ -77,8 +77,8 @@ impl ContainerLog {
             self.drivers
                 .iter_mut()
                 .map(|x| match x {
-                    LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.reopen(),
-                    LogDriver::JsonInterface(ref mut json_logger) => json_logger.reopen(),
+                    LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.reopen().boxed(),
+                    LogDriver::JsonInterface(ref mut json_logger) => json_logger.reopen().boxed(),
                 })
                 .collect::<Vec<_>>(),
         )
```

File: conmon-rs/server/src/container_log.rs [hunk 3]
```diff
@@ -61,14 +71,14 @@ impl ContainerLog {
...
-
-    /// Reopen the container logs.
+      /// Reopen the container logs.
     pub async fn reopen(&mut self) -> Result<()> {
         join_all(
             self.drivers
                 .iter_mut()
                 .map(|x| match x {
                     LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.reopen(),
+                    LogDriver::Json(ref mut json_logger) => json_logger.reopen(),
```
Comment (wasup-yash) [inline conmon-rs/server/src/container_log.rs:81]: hey @saschagrunert  I am having a mismatched type error showing here ``match` arms have incompatible types
  --> conmon-rs/server/src/container_log.rs:81:61
   |
79 |                   .map(|x| match x {
   |  __________________________-
80 | |                     LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => cri_logger.reopen(),
   | |                                                                                 ------------------- this is found to be of type `impl futures::Future<Output = std::result::Result<(), anyhow::Error>>`
81 | |                     LogDriver::Json(ref mut json_logger) => json_logger.reopen(),
   | |                                                             ^^^^^^^^^^^^^^^^^^^^ expected future, found a different future
82 | |                 })
   | |_________________- `match` arms have incompatible types` 
   and `await` on both do not remove the the type error (as the compiler suggests) , so what needs to be change in order to implement the type or should there be change in resullt type of crilogger ? (idk)

Comment (haircommander) [reply conmon-rs/server/src/container_log.rs:81]: did https://github.com/containers/conmon-rs/pull/1744#pullrequestreview-1661525911 change the error output at all?

**Comment by saschagrunert** [comment]
May I ask you to rebase on top of the latest `main` branch to resolve the conflict in `conmon-rs/server/Cargo.toml`?

**Comment by wasup-yash** [comment]
> May I ask you to rebase on top of the latest `main` branch to resolve the conflict in `conmon-rs/server/Cargo.toml`?

Hey can you check now, as only one check is failing which is `ci/lint-rustfmt(pull_request)`

**Comment by saschagrunert** [comment]
@wasup-yash can you apply `cargo fmt` to the PR to format the sources?

**Comment by wasup-yash** [comment]
> @wasup-yash can you apply `cargo fmt` to the PR to format the sources?

done

**Comment by wasup-yash** [comment]
> @wasup-yash can you apply `cargo fmt` to the PR to format the sources?

How do i resolve these failing tests now? ` which is 
ci / test-integration (pull_request) ` and `ci / go-lint (pull_request) `

**Comment by saschagrunert** [comment]
Can you try `golangci-lint run --fix` and commit the changes?

**Comment by saschagrunert** [comment]
Now we're at: 
```
Warning: var-naming: const LogDriverTypeJsonLogger should be LogDriverTypeJSONLogger (revive)
```

**Comment by wasup-yash** [comment]
> Now we're at:
> 
> ```
> Warning: var-naming: const LogDriverTypeJsonLogger should be LogDriverTypeJSONLogger (revive)
> ```

yup updated that only maybe it'll fix

File: conmon-rs/server/src/json_logger.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                    self.bytes_written = 0;
+                }
+            }
+
+            let file = self.file.as_mut().context(Self::ERR_UNINITIALIZED)?;
+            file.write_all(bytes).await?;
+            file.write_all(b"\n").await?; // Newline for each JSON log entry
+
+            // Flush the buffered writer to ensure the log entry is written to the file
+            //file.flush().await.context("Failed to flush log entry")?;
```
Comment (haircommander) [inline conmon-rs/server/src/json_logger.rs:81]: can you drop commented out code

Comment (wasup-yash) [reply conmon-rs/server/src/json_logger.rs:81]: okay

File: conmon-rs/common/proto/conmon.capnp [hunk 5]
```diff
@@ -56,8 +56,9 @@ interface Conmon {
         maxSize @2 :UInt64;
 
         enum Type {
-            # The CRI logger, requires `path` to be set.
+            # The CRI and logger, requires `path` to be set.
```
Comment (saschagrunert) [inline conmon-rs/common/proto/conmon.capnp:59]:
```suggestion
            # The CRI logger, requires `path` to be set.
```

File: conmon-rs/common/proto/conmon.capnp [hunk 6]
```diff
@@ -56,8 +56,9 @@ interface Conmon {
...
             containerRuntimeInterface @0;
+            json @1;
```
Comment (saschagrunert) [inline conmon-rs/common/proto/conmon.capnp:61]:
```suggestion
            # The JSON logger, requires `path` to be set.
            json @1;
```

File: conmon-rs/server/Cargo.toml [hunk 7]
```diff
@@ -31,6 +31,7 @@ prctl = "1.0.0"
 regex = "1.9.6"
 sendfd = { version = "0.4.3", features = ["tokio"] }
 serde = { version = "1.0.188", features = ["derive"] }
+serde_json = "1.0"
```
Comment (saschagrunert) [inline conmon-rs/server/Cargo.toml:34]: Please use the full version here
```suggestion
serde_json = "1.0.107"
```

File: conmon-rs/server/src/container_log.rs [hunk 8]
```diff
@@ -13,36 +13,43 @@ pub struct ContainerLog {
     drivers: Vec<LogDriver>,
 }
 
-#[derive(Debug)]
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:16]: Why does deriving `Debug` not work? It should :thinking:

File: conmon-rs/server/src/container_log.rs [hunk 9]
```diff
@@ -13,36 +13,43 @@ pub struct ContainerLog {
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ContainerRuntimeInterface(arg0) => f
+                .debug_tuple("ContainerRuntimeInterface")
+                .field(arg0)
+                .finish(),
+            Self::Json(arg0) => f.debug_tuple("Json").field(arg0).finish(),
+        }
+    }
 }
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:31]: I don't think we need this, we can just derive debug.

File: conmon-rs/server/src/container_log.rs [hunk 10]
```diff
@@ -13,36 +13,43 @@ pub struct ContainerLog {
...
 
 impl ContainerLog {
-    /// Create a new default SharedContainerLog.
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:22]: Let's keep this comment.

File: conmon-rs/server/src/container_log.rs (hunk 8 above)
Comment (wasup-yash) [reply conmon-rs/server/src/container_log.rs:16]: It didn't worked as the code changed cuz of merge conflicts, needs to look and use it as it did before...

File: conmon-rs/server/src/container_log.rs [hunk 11]
```diff
@@ -78,24 +89,35 @@ impl ContainerLog {
         Ok(())
     }
 
-    /// Write the contents of the provided reader into all loggers.
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:81]: Let's keep those comments.

File: conmon-rs/server/src/container_log.rs [hunk 12]
```diff
@@ -12,47 +12,56 @@ pub type SharedContainerLog = Arc<RwLock<ContainerLog>>;
...
-
 #[derive(Debug)]
 enum LogDriver {
     ContainerRuntimeInterface(CriLogger),
+    Json(JsonLogger),
 }
 
 impl ContainerLog {
-    /// Create a new default SharedContainerLog.
+    //Create a new default SharedContainerLog
```
Comment (saschagrunert) [inline conmon-rs/server/src/container_log.rs:22]: Those should be doc comments, please revert your change.
```suggestion
    /// Create a new default SharedContainerLog
```

File: .vscode/settings.json [hunk 13]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (haircommander) [inline .vscode/settings.json:1]: can we add `.vscode` to the `.gitignore`?

File: conmon-rs/server/src/container_log.rs (hunk 10 above)
Comment (haircommander) [reply conmon-rs/server/src/container_log.rs:22]:
```
/// Create a new default SharedContainerLog.
```
can it be put back to this so it's registered as a doc for the func

File: pkg/client/client_test.go [hunk 14]
```diff
@@ -672,3 +672,72 @@ var _ = Describe("ConmonClient", func() {
...
+					}},
+				})
+				Expect(err).NotTo(Succeed())
+				Expect(err.Error()).To(ContainSubstring(`executable file not found in $PATH"`))
+			})
+			It(testName("should log in JSON format for LogDriverTypeJSONLogger", terminal), func() {
+				tr = newTestRunner()
+				tr.createRuntimeConfigWithProcessArgs(terminal, []string{"invalid"}, nil)
+				sut = tr.configGivenEnv()
+				_, err := sut.CreateContainer(context.Background(), &client.CreateContainerConfig{
```
Comment (haircommander) [inline pkg/client/client_test.go:720]: didn't handle error here

File: pkg/client/client_test.go [hunk 15]
```diff
@@ -672,3 +672,72 @@ var _ = Describe("ConmonClient", func() {
...
+				// This is a basic check assuming that a valid JSON starts and ends with braces '{' and '}'
+				Expect(strings.TrimSpace(string(logContent))).To(SatisfyAll(
+					Not(BeEmpty()),
+					HavePrefix("{"),
+					HaveSuffix("}"),
+				))
+			})
+		}		
+	})
+	})
```
Comment (haircommander) [inline pkg/client/client_test.go:741]: indentation is a bit off here

File: conmon-rs/server/src/container_log.rs (hunk 12 above)
Comment (wasup-yash) [reply conmon-rs/server/src/container_log.rs:22]: ok

File: pkg/client/client_test.go [hunk 16]
```diff
@@ -320,7 +321,7 @@ var _ = Describe("ConmonClient", func() {
 				rssAfter := vmRSSGivenPID(pid)
 				GinkgoWriter.Printf("VmRSS after: %d\n", rssAfter)
 				GinkgoWriter.Printf("VmRSS diff: %d\n", rssAfter-rssBefore)
-				Expect(rssAfter - rssBefore).To(BeNumerically("<", 2000))
+				Expect(rssAfter - rssBefore).To(BeNumerically("<", 10000))
```
Comment (saschagrunert) [inline pkg/client/client_test.go:324]: I think `serde_json` eats memory.

**Comment by saschagrunert** [comment]
@haircommander PTAL

**Pull request merged**